src/lib/lossless_cast.hpp

```diff
@@ -0,0 +1,229 @@
+#pragma once
+
+#include <optional>
+#include <type_traits>
+
+#include "all_type_variant.hpp"
+#include "null_value.hpp"
+
+/**
+ * This file defines the following casting functions that guarantee **lossless** conversion between data types and
+ * return std::nullopt if no lossless conversion was possible.
+ *      lossless_cast<Target>(Source)
+ *      lossless_variant_cast<Target>(AllTypeVariant)
+ *      lossless_variant_cast(Source, DataType)
+ *
+ * "Lossless" means just that: No information must be lost during casting. Typical lossful operations would be
+ *      - float-to-int casts with the float having a non-zero fractional part
+ *      - long-to-int casts where the long was > MAX_INT
+ *      - string-to-int where the string contained no number, or a number > MAX_INT.
+ *      - double-to-float where the significand of the double has bits set that cannot be represented in float
+ *
+ * Whenever possible you should use these casts over "comfortable" casts such as static_cast or boost::lexical_cast.
+ * Those might incur information loss and might thus introduce bugs. See #1306, #1525.
+ *
+ * THINK ABOUT AN ALGORITHM'S DATA, THE POSSIBLE STATES IT MIGHT HAVE, AND HANDLE THEM PROPERLY.
+ */
+
+namespace opossum {
+
+inline std::tuple<bool, int32_t, uint32_t> decompose_floating_point(float f) {
```

Comment:
1. Why is this necessary?
2. "The value representation of floating-point types is implementation-defined" - This should be at least documented and statically(?) asserted.

Reply:
1. To check whether the float can be losslessly converted to an integer. I have searched a bit and found - to my surprise and anger - no premade function that does this. 

TBH, the float->int conversion is the shakiest part of this PR and requires more thought and testing.

Can you do `f <= INT_MAX && f >= INT_MIN` followed by a `static_cast<int>(f)`? `f <= INT_MAX` will implicitly cast INT_MAX to float to perform the comparison, as I understand it. If roundup happens during this conversion `f <= INT_MAX` will return true for some floats that are larger than INT_MAX. Actually looking at the bits seemed the best way around this.

I COULD also study IEEE754 with respect to rounding from int to float, maybe `f <= INT_MAX && f >= INT_MIN` is guaranteed to work. 

2. Added `static_assert(std::numeric_limits<float>::is_iec559, "IEEE 754 floating point representation expected.");`

Reply:
I haven't had a coffee yet, so this might be stupid but why don't we cast the number in question into the other type, cast it back, and see if it changed?

Reply:
I did it like that before. You'll get UBSan, rightfully, complaining about casts from floats >INT_MAX to int, e.g. See my previous answer why I am hesitant to check for f > INT_MAX.

Reply:
Can't you exclude that case by comparing it to `< (float)(INT_MAX)` first?

Reply:
After some more thought I couldn't find a case where a check `f < (float)INT_MAX` yields to false positives/negatives do to conversion. The float-decomposition, thus, was removed. Much cleaner.

Reply:
Nope, doesn't work, here is why:

```c++
#include <limits>
#include <iostream>

int main(int argc, char **argv) {
  // INT_MAX is 2'147'483'647

  std::cout << "INT_MAX is " << std::numeric_limits<int32_t>::max() << std::endl;
  std::cout << "INT_MAX casted to float is " << std::fixed << static_cast<float>(std::numeric_limits<int32_t>::max()) << std::endl;
  std::cout << "Thus 2'147'483'648.0f, which should be greater than INT_MAX, is actually not greater: " << (2'147'483'648.0f > std::numeric_limits<int32_t>::max()) << std::endl;
  std::cout << "So any lossless_cast algorithm based on comparison would think it can convert 2'147'483'648.0f to a 32bit int, but it actually cannot." << std::endl;

}
```

2'147'483'648.0f, which is greater than INT_MAX, would not be detected as such.

Reply:
Mmh, how about checking if `modf` returns `0.0`?

Reply:
that would still not tell us whether the float is bigger than INT_MAX and thus doesn't tell us whether the `static_cast<int>(some_float)` is safe...

```c++
  auto input_float = 2'147'483'648.0f;

  float integral;
  auto fractional = std::modf(input_float, &integral);

  std::cout << fractional << std::endl;
  std::cout << integral << std::endl;
  std::cout << (integral > std::numeric_limits<int32_t>::max()) << std::endl;
  std::cout << (input_float > std::numeric_limits<int32_t>::max()) << std::endl;
```

gives 

```
0
2.14748e+09
0
0
```

Reply:
Haha, well - the problem is that `std::numeric_limits<int32_t>::max()` cannot be expressed as a float. `std::numeric_limits<int32_t>::max() + 1`, however, can:

```
#include <cmath>
#include <limits>
#include <iostream>

bool is_int(float f) {
	float integral_part;
	if(std::modf(f, &integral_part) != 0.0) return false;
	if(integral_part >= (float)0x80000000u) return false;
	// TODO handle the negative side here
	return true;
}

int main(int argc, char **argv) {
  for (auto f : {5.0f, 5.5f, 2'147'483'647.0f, 2'147'483'648.0f, std::nexttoward(2'147'483'648.0f, 0.f)}) {
  	std::cout << std::fixed << f << " " << is_int(f) << std::endl;
  }
}
```

```
5.000000 1
5.500000 0
2147483648.000000 0
2147483648.000000 0
2147483520.000000 1
```

Reply:
> the problem is that std::numeric_limits<int32_t>::max() cannot be expressed as a float

Yes, that's what I've been trying to express the entire time - I thought the whole discussion in this thread and to some degree the whole purpose of this PR is just that: There are floats that aren't representable as ints, there are ints that aren't representable as float...

Reply:
Sure, I'm just trying to find a way around having to play with the bit representation of the float. Maybe it's easier to talk about this in person.

Reply:
Let me just add one more thought though, lest I forget:

> There are floats that aren't representable as ints

Yes, but those are all outside of [INT_MIN, INT_MAX]. Checking if the float is within those boundaries and has no fractional part should thus be sufficient.

Reply:
What do you think of this?

```c++
// Floating Point Type to 32/64 bit Integral Type
template <typename Target, typename Source>
std::enable_if_t<std::is_floating_point_v<Source> && (std::is_same_v<Target, int32_t> || std::is_same_v<Target, int64_t>), std::optional<Target>>
lossless_cast(const Source& source) {
  static_assert(std::numeric_limits<float>::is_iec559, "IEEE 754 floating point representation expected.");

  auto integral_part = Source{};

  // No lossless float-to-int conversion possible if the source float has a fractional part
  if (std::modf(source, &integral_part) != 0.0) {
    return std::nullopt;
  }

  // Check the floating point value against explicitly defined boundary values. The boundary values are the
  // the closest-to-zero floating point values that are NOT representable in the given integral type.
  // (Note: We evaluated multiple approaches to identify whether a float/double is representable as a int32/64. None 
  //        of them was really portable/readable/"clean". The bounding values approach at least makes intuitive sense.) 
  if constexpr (std::is_same_v<Source, float> && std::is_same_v<Target, int32_t>) {
    if (source >= 2'147'483'648.0f || source <= -2'147'483'904.0f) return std::nullopt;
  } else if constexpr (std::is_same_v<Source, double> && std::is_same_v<Target, int32_t>) {
    if (source >= 2'147'483'648.0 || source <= -2'147'483'649.0) return std::nullopt;
  } else if constexpr (std::is_same_v<Source, float> && std::is_same_v<Target, int64_t>) {
    if (source >= 9'223'372'036'854'775'808.0f || source <= -9'223'373'136'366'403'584.0f) return std::nullopt;
  } else if constexpr (std::is_same_v<Source, double> && std::is_same_v<Target, int64_t>) {
    if (source >= 9'223'372'036'854'775'808.0 || source <= -9'223'372'036'854'777'856.0) return std::nullopt;
  }

  return static_cast<Target>(source);
}
```